### DISCUSSION TITLE
Build process change WRT references

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#6997] [Q&A] [answered]
Can somebody point me to where I would ask/request for a change to how projects are built? 
Specifically, I'd like it to check for the referenced DLL's public surface area, and only treat it as a reason to build _this_ project if the referenced DLL's public surface area changed, and not whenever the referenced DLL was touched in any way (such as due to a change in a private member or method body code).
This would drastically reduce the build times for large solutions, because if there is a dependency chain of A->B->C->D, and D gets a non-public change, then currently, we rebuild A, B, C, and D; whereas with my proposal, we would only build D, run the public surface area comparison on C (which should be a simple byte array comparison and therefore very quick), and (granted D's public surface area didn't change), skip building C, and do absolutely nothing for A and B.
Also, deployments would be faster, because the build tools can optimize and only copy the few DLLs that actually need to be copied.
And finally, this behavior can be made optional, so those users who prefer the old behavior (if they feel it's more reliable) can continue using it.

**Comment by HaloFour** [comment, marked as answer]
Sounds like a build system concern, you could try with [MSBuild](https://github.com/dotnet/msbuild).

**Reply by RikkiGibson** [reply]
I think an optimization already exists which builds a *ref assembly* (which includes only the externally accessible surface area of the assembly), and simply looks at whether the ref assembly changed in order to decide whether to rebuild downstream projects. I may be remembering wrong, though.

**Reply by TahirAhmadov** [reply]
That does seem to be the case, however only the most recent version of VS (17.5) has the full suite of features to achieve what I need, and I'm going to try it soon.

https://github.com/dotnet/msbuild/discussions/8495

**Comment by erik-kallen** [comment]
I believe Rider has a setting that does this.
